draft-barnes-mls-protocol.md

```diff
@@ -1435,44 +1434,128 @@ all arrive at the following state:
+is RECOMMENDED for this secret to be updated with every message.
+Since Application AEAD keys are automatically updated at each Group operation,
+it is not expected for the Clients to be very agressive on independant updates.
+In all cases, participant MUST refresh their Application secret before
+exhaustion of the associated AEAD keys.
+
+Note that each change to the Group through a Handshake message will cause
+a change of the Group Secret. Hence this change MUST be applied before encrypting
+any new Application message. This is required for obvious confidentiality reasons
+regarding who can encrypt and decrypt Application messages.
+
+[[OPEN ISSUE: This Key Schedule being the one of the most important part of the protocol
+its security should be formalized and be as close as possible of the TLS 1.3 Key Schedule.]]
+
+[[OPEN ISSUE: Should we maintain a Transcript hash of the Application Data ?
+It would be surprising if the message counter was enough.]]
+
+## Updating the Application Secret
+
+As each Application Secret MUST be updated to provide Application messages
+with better FS and PCS guarantees. The next generation of Application Secret
+is computed by generating an Application_Secret_N+1 from Application_Secret_N
+as described in this section then re-deriving the traffic keys as described in
+the previous section.
+
+The next-generation Application_Secret_N+1 can happen independantly from an update
+triggered by a Group Secret change and is computed as:
+
+~~~~
+Application_Secret_N+1 =
+    HKDF-Expand-Label(Application_Secret_N,"mls app upd","", Hash.length)
+~~~~
+
+Once Application_Secret_N+1 and its associated traffic keys have
+been computed, implementations SHOULD delete Application_Secret_N and
+its associated traffic keys as soon as possible. For MLS, the usability
+probably requires to keep the Application secrets and keys for a certain
+amount of time to retain the ability to decrypt messages possibly in transit
+while the updating being done. Note that keeping these secrets will considerably
+weaken the cryptographic security guarantees expected at the protocol level.
+
+## Application AEAD Key Calculation
+
+The Application AEAD keying material is generated from the following
+input values:
+
+- The Application Secret value
+- A purpose value indicating the specific value being generated and by who.
+- The length of the bytestream being generated
+
+The traffic keying material is generated from an input traffic secret value using:
+
+~~~~
+[sender]_write_key =
+    HKDF-Expand-Label(Application Secret,[sender]+"key","",key_length)
+
+[sender]_write_iv  =
+    HKDF-Expand-Label(Application Secret,[sender]+"iv","",iv_length)
+~~~~
+
+\[sender] denotes the identifier of the Group participant and + the
+concatenation operator.
+
+Note that since, the Application keys are dependant on the identity of
+the sender, the label caracterizing this identity MUST be part of the
+Associated-Data when performing AEAD encryption.
```

Comment:
If you're going to say "MUST", you have to say how :)

And in any case, is that necessary?  The identity of the sender is included in the key derivation.